tests: logging/dictionary: fix pytest failures on hardware

dcpleung opened the pull request:
These should fix pytest failures on hardware.

Tested on `frdm_k64f` and `up_squared`.

Fixes #75253

dcpleung committed:
tests: logging/dictionary: fix pytest log parsing regex

When doing device testing, there may be some other characters
being printed before the dictionary logging header string.
Also the delay boot banner may also be there. So change
the regex to ignore those them.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/dictionary: fix pytest when built with sysbuild

With sysbuild, the test app is actually built under another
layer of directory where the dictionary JSON file will reside.
Instead of simply using dut.device_config.build_dir, use
dut.devive_config.app_build_dir instead, which takes into
account of that.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

gchwier commented on tests/subsys/logging/dictionary/pytest/test_logging_dictionary.py:
```diff
@@ -33,13 +33,21 @@ def process_logs(dut: DeviceAdapter, build_dir):
+        dictionary_json = os.path.join(build_dir, "zephyr", "dictionary", "log_dictionary.json")
+    else:
+        dictionary_json = os.path.join(build_dir, "zephyr", "log_dictionary.json")
+
```
look into https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/pytest-twister-harness/src/twister_harness/twister_harness_config.py#L42

you can use as `dut.device_config.app_build_dir` in `test_logging_dictionary` to have both sysbuild and non-sysbuild scenarios covered

dcpleung replied:
That's good to know. Thanks.

dcpleung committed:
tests: logging/dictionary: wait longer for first output

Some hardware takes longer to boot before any output is visible
from the app. So lengthen the initial timeout on waiting for
any output.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

katgiadla commented on tests/subsys/logging/dictionary/pytest/test_logging_dictionary.py:
```diff
@@ -38,8 +38,10 @@ def process_logs(dut: DeviceAdapter, build_dir):
+    # as the log parser requires file as input.
+    # Timeout is intentionally long. Twister will
+    # timeout earlier with per-test timeout.
+    handler_output = dut.readlines_until(regex = '.*##ZLOGV1##[0-9]+', timeout = 600.0)
```
is this timeout too long (this test could executes ca. 10 min after changes)?

dcpleung replied:
Twister has its own timeout outside of this which is shorter, and will terminate the test accordingly. So a large number here doesn't change that aspect of twister.

katgiadla reviewed (approved):
Works on Nordic platforms.
Thanks
